File: 08_gpu_teams/problems/02_gpu_integral/gpu_integral.cpp
#include <cstdio>
#include <cstdlib>
#include <cmath>

int main(int argc, char ** argv) {
  long n = (argc > 1 ? atol(argv[1]) : 100000000L);
  double dx = 1.0 / (double)n;
  double s = 0.0;

  /* 中点則で ∫_0^1 4/(1+x^2) dx = π を GPU 上で計算する.
     s は総和 (スカラ) なので reduction(+:s) を使う.
     スカラはコンパイラが自動的に転送するので map は不要. */
  // TODO: GPU上で reduction(+:s) を使って総和を求め π を計算せよ.

  double pi = s * dx;
  printf("n = %ld\n", n);
  printf("pi  = %.15f\n", pi);
  printf("M_PI = %.15f\n", M_PI);
  printf("error = %.3e\n", fabs(pi - M_PI));
  return 0;
}

File: 08_gpu_teams/problems/02_gpu_integral/gpu_integral.f90
program gpu_integral
  character(len=32) :: arg
  integer(8) :: n, i
  real(8) :: dx, s, x, pi
  real(8), parameter :: pi_ref = 3.141592653589793d0
  n = 100000000_8
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) n
  end if
  dx = 1.0d0 / dble(n)
  s = 0.0d0

  ! 中点則で ∫_0^1 4/(1+x^2) dx = π を GPU 上で計算する.
  ! s は総和 (スカラ) なので reduction(+:s) を使う.
  ! スカラはコンパイラが自動的に転送するので map は不要.
  ! TODO: GPU上で reduction(+:s) を使って総和を求め π を計算せよ.

  pi = s * dx
  print "(a,i0)", "n = ", n
  print "(a,f0.15)", "pi  = ", pi
  print "(a,f0.15)", "M_PI = ", pi_ref
  print "(a,es10.3)", "error = ", abs(pi - pi_ref)
end program gpu_integral
